### PR TITLE
Feature/sts business rule additions

### PR DESCRIPTION
**Pull request by hbgwhite**
Added validations for Group IDs and English language test forms

### CODE REVIEW

File: ProcessSmarterTestPackage/PostProcessors/FormPartitionPostProcessor.cs [hunk 1]
```diff
@@ -16,23 +16,101 @@ public override IList<ValidationError> GenerateErrors()
...
             var partitionIdentifier = Processor.ChildNodeWithName("identifier").ValueForAttribute("uniqueid");
             var itemGroups = Processor.ChildNodesWithName("itemgroup").ToList();
+            const string groupIdPattern = @"(\d+-\d+):([G|I])-(\d+-\d+)";
+
             foreach (var itemGroup in itemGroups)
             {
-                if (!itemGroup.ChildNodeWithName("identifier")
-                    .ValueForAttribute("uniqueid").Split(':').First()
-                    .Equals(partitionIdentifier, StringComparison.OrdinalIgnoreCase))
+                var match = Regex.Match(itemGroup.ChildNodeWithName("identifier")
```
Comment (hbgwhite) [inline ProcessSmarterTestPackage/PostProcessors/FormPartitionPostProcessor.cs:32]: The changes here emit errors around group identifiers. In the Student application, groups identifiers whose second portion begins with "G-" are treated differently than those that start with "I-". Other identifier prefixes are not recognized. By convention, the first part of a group ID (the part preceding the colon) should be the form partition ID that the group belongs to. The part following should be either the associated stimuli ID found in the passageref element (for "G-" item groups) or the item ID of the single item in the group (for "I-" groups). Errors emitted as "Severe" will cause the application to crash at runtime when it attempts to load those groups. Benign errors here are violations of the convention for group IDs that have no material effect on loading and administering assessments.

File: ProcessSmarterTestPackage/PostProcessors/TestSpecificationPostProcessor.cs [hunk 2]
```diff
@@ -131,7 +131,25 @@ public override IList<ValidationError> GenerateErrors()
                     .ChildNodeWithName("itempool")
                     .ChildNodesWithName("testitem")));
 
-            var testForms = Processor.ChildNodeWithName(PackageType.ToString()).ChildNodesWithName("testform");
+            var testForms = Processor.ChildNodeWithName(PackageType.ToString()).ChildNodesWithName("testform").ToList();
```
Comment (hbgwhite) [inline ProcessSmarterTestPackage/PostProcessors/TestSpecificationPostProcessor.cs:134]: TDS looks for an English test form at execution time, even if the assessment is being administered in a different language. The absence of this form will cause the application to crash at runtime. This validation ensures that the package contains at least one English language test form.

File: ProcessSmarterTestPackage/PostProcessors/ItemGroupPostProcessor.cs [hunk 3]
```diff
@@ -14,6 +14,7 @@ public override IList<ValidationError> GenerateErrors()
             var result = new List<ValidationError>();
```
Comment (hbgwhite) [inline ProcessSmarterTestPackage/PostProcessors/ItemGroupPostProcessor.cs:15]: Some minor simplification

File: ProcessSmarterTestPackage/PostProcessors/FormPartitionPostProcessor.cs [hunk 4]
```diff
@@ -16,23 +16,101 @@ public override IList<ValidationError> GenerateErrors()
...
+                                ErrorSeverity = ErrorSeverity.Benign,
+                                GeneratedMessage =
+                                    $"[Itemgroup identifier {itemGroup.ChildNodeWithName("identifier").ValueForAttribute("uniqueid")} is a single-item specified \"I-\" item group, but does not have a reference to a single child groupitem element]",
+                                Key = "uniqueid",
+                                ItemId = itemGroup.ChildNodeWithName("identifier").ValueForAttribute("uniqueid"),
+                                PackageType = PackageType,
+                                Location = "itemgroup/identifier"
+                            });
+                        }
+                        else if (!match.Groups[3].Value.Equals(
```
Comment (hbgwhite) [inline ProcessSmarterTestPackage/PostProcessors/FormPartitionPostProcessor.cs]: "I-" item group identifiers second portion should match their group item by convention.

File: ProcessSmarterTestPackage/PostProcessors/FormPartitionPostProcessor.cs [hunk 5]
```diff
@@ -16,23 +16,101 @@ public override IList<ValidationError> GenerateErrors()
...
+                                Key = "uniqueid",
+                                ItemId = itemGroup.ChildNodeWithName("identifier").ValueForAttribute("uniqueid"),
+                                PackageType = PackageType,
+                                Location = "itemgroup/identifier"
+                            });
+                        }
+                    }
+                    else if (match.Groups[2].Value.Equals("I"))
+                    {
+                        if (itemGroup.ChildNodesWithName("groupitem").Count() != 1)
```
Comment (hbgwhite) [inline ProcessSmarterTestPackage/PostProcessors/FormPartitionPostProcessor.cs]: "I-" item group identifiers indicate that the group in question should only contain a single item

File: ProcessSmarterTestPackage/PostProcessors/FormPartitionPostProcessor.cs [hunk 6]
```diff
@@ -16,23 +16,101 @@ public override IList<ValidationError> GenerateErrors()
...
+                                Key = "uniqueid",
+                                ItemId = itemGroup.ChildNodeWithName("identifier").ValueForAttribute("uniqueid"),
+                                PackageType = PackageType,
+                                Location = "itemgroup/identifier"
+                            });
+                        }
+                        else if (!match.Groups[3].Value.Equals(
+                            itemGroup.ChildNodeWithName("passageref").ValidatedAttributes["passageref"].Value))
+                        {
+                            result.Add(new ValidationError
```
Comment (hbgwhite) [inline ProcessSmarterTestPackage/PostProcessors/FormPartitionPostProcessor.cs]: "G-" item group identifiers second portion should match their associated passage ID by convention.

File: ProcessSmarterTestPackage/PostProcessors/FormPartitionPostProcessor.cs [hunk 7]
```diff
@@ -16,23 +16,101 @@ public override IList<ValidationError> GenerateErrors()
...
+                                $"[Itemgroup identifier {itemGroup.ChildNodeWithName("identifier").ValueForAttribute("uniqueid")} first portion {match.Groups[1]} does not match partition ID and expected value {partitionIdentifier}]",
+                            Key = "uniqueid",
+                            ItemId = itemGroup.ChildNodeWithName("identifier").ValueForAttribute("uniqueid"),
+                            PackageType = PackageType,
+                            Location = "itemgroup/identifier"
+                        });
+                    }
+                    if (match.Groups[2].Value.Equals("G"))
+                    {
+                        if (itemGroup.ChildNodesWithName("passageref").Count() != 1)
```
Comment (hbgwhite) [inline ProcessSmarterTestPackage/PostProcessors/FormPartitionPostProcessor.cs]: "G-" items that do not contain a passageref that matches a valid stimuli fail at runtime.

File: ProcessSmarterTestPackage/PostProcessors/FormPartitionPostProcessor.cs [hunk 8]
```diff
@@ -16,23 +16,101 @@ public override IList<ValidationError> GenerateErrors()
...
-                    .ValueForAttribute("uniqueid").Split(':').First()
-                    .Equals(partitionIdentifier, StringComparison.OrdinalIgnoreCase))
+                var match = Regex.Match(itemGroup.ChildNodeWithName("identifier")
+                    .ValueForAttribute("uniqueid"), groupIdPattern);
+                if (!match.Success)
                 {
                     result.Add(new ValidationError
                     {
-                        ErrorSeverity = ErrorSeverity.Benign,
+                        ErrorSeverity = ErrorSeverity.Severe,
```
Comment (hbgwhite) [inline ProcessSmarterTestPackage/PostProcessors/FormPartitionPostProcessor.cs:39]: These identifiers are parsed when retrieving items to present in Student. If they do not follow this pattern, they may cause (hard to debug) failures at runtime.

File: ProcessSmarterTestPackage/PostProcessors/FormPartitionPostProcessor.cs (hunk 1 above)
Comment (jtreuting) [reply ProcessSmarterTestPackage/PostProcessors/FormPartitionPostProcessor.cs:32]: Got it, makes sense.

**Review by jtreuting — APPROVED** [review]
Can you add inline comments on the new validation checks that explain what they are doing. The checks aren;'t easy to understand without that context.  Other than that looks good.

**Pull request merged**